x-pack/plugins/upgrade_assistant/server/lib/reindexing/index_settings.ts

```diff
@@ -0,0 +1,140 @@
+ *    For the mapping type:
+ *    ```
+ *      {
+ *        "field1": { "type": "boolean" },
+ *        "nested": {
+ *          "field2": { "type": "boolean" }
+ *        }
+ *      }
+ *    ```
+ *    The fieldPaths would be: `[['field1'], ['nested', 'field2']]`
+ * @param properties
+ */
+export const findBooleanFields = (properties: MappingProperties): string[][] =>
+  Object.keys(properties).reduce(
+    (res, propertyName) => {
+      if (properties[propertyName].type === 'boolean') {
+        // If this field is a boolean, add it
+        res.push([propertyName]);
+      } else if (properties[propertyName].properties) {
+        // If this is a nested object/array get the nested fields and prepend the field path with the current field.
+        const nested = findBooleanFields(properties[propertyName].properties!);
+        res = [...res, ...nested.map(n => [propertyName, ...n])];
+      }
+
+      return res;
+    },
+    [] as string[][]
+  );
+
+const removeUnsettableSettings = (settings: FlatSettings['settings']) =>
+  omit(settings, [
+    'index.uuid',
+    'index.blocks.write',
+    'index.creation_date',
+    'index.legacy',
+    'index.mapping.single_type',
+    'index.provided_name',
+    'index.routing.allocation.initial_recovery._id',
+    'index.version.created',
+    'index.version.upgraded',
+  ]);
+
+const updateFixableSettings = (settings: FlatSettings['settings']) => {
+  const delayedTimeout = settings['index.unassigned.node_left.delayed_timeout'];
+  if (delayedTimeout && parseInt(delayedTimeout, 10) < 0) {
+    settings['index.unassigned.node_left.delayed_timeout'] = '0';
+  }
+
+  return settings;
+};
+
+const validateSettings = (settings: FlatSettings['settings']) => {
+  if (settings['index.shard.check_on_startup'] === 'fix') {
+    throw new Error(`index.shard.check_on_startup cannot be set to 'fix'`);
+  }
+
+  if (settings['index.percolator.map_unmapped_fields_as_string']) {
+    throw new Error(`index.percolator.map_unmapped_fields_as_string is no longer supported.`);
+  }
+
+  return settings;
+};
+
+// Use `flow` to pipe the settings through each function.
+const transformSettings = flow(
+  removeUnsettableSettings,
+  updateFixableSettings,
+  validateSettings
+);
+
+const updateFixableMappings = (mappings: FlatSettings['mappings']) => {
+  if (mappings._default_) {
+    delete mappings._default_;
+  }
+
+  const mapping = getSingleMappingType(mappings);
+  if (mapping && mapping._all) {
+    delete mapping._all;
+  }
+
+  return mappings;
+};
+
+const transformMappings = flow(updateFixableMappings);
+
+export const getSingleMappingType = (
+  mappings: FlatSettings['mappings']
+): TypeMapping | undefined => {
+  const mappingTypes = Object.keys(mappings)
+    // Ignore _default_ mapping types.
+    .filter(t => t !== '_default_');
+
+  if (mappingTypes.length > 1) {
+    throw new Error(`Indices with more than one mapping type are not supported in 7.0.`);
+  }
+
+  return mappings[mappingTypes[0]];
```

Comment:
Fairly certain we should be using `_doc` for here.

Reply:
Should be easy enough, I can create a follow up PR for 7.0 after this gets merged.